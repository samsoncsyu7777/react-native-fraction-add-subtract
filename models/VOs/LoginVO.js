class LoginVO {
  constructor() {
    this.action = "login",
    this.operator = "",
    this.passkey = "";
  }    
}

export default LoginVO;