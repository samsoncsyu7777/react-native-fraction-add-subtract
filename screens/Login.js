import React, { useState, useContext, useEffect } from "react";
import {
  widthPercentageToDP as wp2dp,
  heightPercentageToDP as hp2dp,
} from 'react-native-responsive-screen';
import {
  StyleSheet,
  Text,
  View,
  Platform
} from "react-native";
import theme from "../Theme";
import Icon from 'react-native-vector-icons/EvilIcons';
import {
  withTheme,
  TextInput,
  ActivityIndicator,
  Badge,
  Button,
  Paragraph,
  Dialog,
  Portal,
  ProgressBar,
  Colors
} from 'react-native-paper';
import Inventory from "./Inventory";
import Dashboard from "./Dashboard";
import axios from 'axios';

import TokenRequestVO from "../models/VOs/TokenRequestVO";
import TokenResponseDTO from "../models/DTOs/TokenResponseDTO";
import LoginVO from "../models/VOs/LoginVO";
import secret from "../Secret";
import { UserContext } from "../contexts/UserContext";

function Login(props) {
  const { user, setUser } = useContext(UserContext);
  const [clientId, setClientId] = useState('');
  const [password, setPassword] = useState('');
  const [hiddenPassword, setHiddenPassword] = useState('');
  const [visible, setVisible] = useState(false);
  var loginVO = new LoginVO();
  let tokenResponseDTO = new TokenResponseDTO();
  let hiddenPasswordFull = '*******************************';

  useEffect(() => {
    token();
  });

  const submit = (e) => {
    e.preventDefault();
    loginVO.operator = clientId;
    loginVO.passkey = password;
    signIn();

  };

  const hideDialog = () => {
    setVisible(false);
    props.navigation.navigate('Dashboard');
  }

  const token = () => {
    let tokenRequestVO = new TokenRequestVO();
    const PROXY_URL = (Platform.OS === 'web') ? 'https://cors-anywhere.herokuapp.com/' : '';
    const URL = secret.URL + '/api/v1/token';

    axios.post(PROXY_URL + URL, tokenRequestVO)
      .then(response => {
        tokenResponseDTO = response.data;
      })
      .catch(e => console.error("Error occured! ", e));
  }

  const signIn = () => {
    const PROXY_URL = (Platform.OS === 'web') ? 'https://cors-anywhere.herokuapp.com/' : '';
    const URL = secret.URL + '/api/v1/user/access';

    const res = fetch(PROXY_URL + URL, {
      method: "post",
      headers: {
        "Content-Type": "application/json",
        "Authorization": "Bearer " + tokenResponseDTO.token
      },
      body: JSON.stringify(loginVO),
    })
      .then((response) => response.json())
      .then((responseJson) => {
        console.log(responseJson);
        if ("token" in responseJson) {
          user.clientId = loginVO.operator;
          user.password = loginVO.passkey;
          user.token = responseJson.token;
          setVisible(true);
        } else if ("message" in responseJson) {
          setMessage(responseJson.message);
          setSeverity("error");
        } else if ("errors" in responseJson) {
          setMessage(
            responseJson.errors[0].param + ": " + responseJson.errors[0].msg
          );
        }
      })
      .catch((error) => {
        console.error(error);
      });
  }

  return (
    <View style={styles.container}>
      <ProgressBar
        progress={0.5}
        color={Colors.red800}
        style={{ height: 10, width: 300, borderRadius: 10 }}
        indeterminate />
      <Text style={styles.text}>CLIENT ID</Text>
      <TextInput
        mode="outlined"
        value={clientId}
        label="CLIENT ID"
        style={styles.textInput}
        onChangeText={clientId => setClientId(clientId)}
      />
      <Badge
        style={styles.badge}
      >
        {clientId.length}
      </Badge>
      <Text style={styles.text}>PASSWORD</Text>
      <TextInput
        mode="outlined"
        label="PASSWORD"
        value={hiddenPassword}
        style={styles.textInput}
        onChangeText={hiddenPassword => {
          var tmp = '';
          if (hiddenPassword.length >= password.length) {
            tmp = password + hiddenPassword[hiddenPassword.length - 1];
          } else {
            tmp = password.slice(0, -1);
          }
          setHiddenPassword(hiddenPasswordFull.substr(0, tmp.length));
          setPassword(tmp);
        }}
      />
      <Badge
        style={styles.badge}
      >
        {password.length}
      </Badge>
      <Icon
        name="arrow-right"
        size={90}
        color="#1690aa"
        onPress={submit}
      />
      <ActivityIndicator
        animating={true}
        color={theme.colors.accent}
        size="large"
      />
      <Portal>
        <Dialog
          visible={visible}
          onDismiss={hideDialog}
          style={styles.dialog}
        >
          <Dialog.Title>Welcome</Dialog.Title>
          <Dialog.Content>
            <Paragraph>Correct Client ID and Password</Paragraph>
          </Dialog.Content>
          <Dialog.Actions>
            <Button onPress={hideDialog}>Done</Button>
          </Dialog.Actions>
        </Dialog>
      </Portal>
    </View>
  );

}

export default withTheme(Login);

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "#e0ffee",
    alignItems: "center",
    justifyContent: "center"
  },

  text: {
    fontSize: hp2dp('4%'),//28,
    fontWeight: "bold",
    color: theme.colors.primary,
    padding: 10
  },

  textInput: {
    width: 300,
    fontSize: hp2dp('3%'),//18,
    fontWeight: "bold",
    backgroundColor: theme.colors.background,
    borderColor: theme.colors.primary,
    borderRadius: 40,
    borderWidth: 2,
    padding: 10,
    marginBottom: 0
  },

  badge: {
    alignSelf: "center",
    width: hp2dp('4%'),//30,
    height: hp2dp('4%'),//30,
    fontSize: hp2dp('3%'),//20,
    paddingTop: hp2dp('1%'),
    marginLeft: 220
  },

  dialog: {
    textAlign: "center",
    backgroundColor: theme.colors.dialog,
    width: wp2dp('60%'),//300,
    alignSelf: "center"
  },

});
