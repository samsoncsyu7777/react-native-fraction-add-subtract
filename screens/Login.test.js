import React from 'react';
import renderer from 'react-test-renderer';
import LoginVO from '../models/VOs/LoginVO';

import Login from "./Login";

describe('<Login />', () => {
  
  it('has 5 children', async () => {
    const tree = renderer.create(<Login />).toJSON();
    expect(tree.children.length).toBe(5);
  });

  it('App renders without crashing', async () => {
    const rendered = renderer.create(<Login />).toJSON();
    expect(rendered).toBeTruthy();
  });

  it('App test against snapshot', async () => {
    const tree = renderer.create(<Login />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('View properties', async () => {
    const tree = renderer.create(<Login />).toJSON();
    expect(tree.children).toHaveLength(5);
    expect(tree.props.style).toHaveProperty('backgroundColor');
  });

});