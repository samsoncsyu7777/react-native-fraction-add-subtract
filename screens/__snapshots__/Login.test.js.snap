// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> App test against snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#e0ffee",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    accessibilityRole="progressbar"
    accessible={true}
    onLayout={[Function]}
  >
    <View
      style={
        Object {
          "backgroundColor": "rgba(198, 40, 40, 0.38)",
          "borderRadius": 10,
          "height": 10,
          "opacity": 0,
          "overflow": "hidden",
          "width": 300,
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "#c62828",
            "flex": 1,
            "transform": Array [
              Object {
                "translateX": -0,
              },
              Object {
                "scaleX": 0.0001,
              },
            ],
            "width": 0,
          }
        }
      />
    </View>
  </View>
  <Text
    style={
      Object {
        "color": "#2080aa",
        "fontSize": 53.5,
        "fontWeight": "bold",
        "padding": 10,
      }
    }
  >
    CLIENT ID
  </Text>
  <View
    style={
      Object {
        "borderColor": "#2080aa",
        "borderRadius": 40,
        "borderWidth": 2,
        "marginBottom": 0,
        "padding": 10,
        "width": 300,
      }
    }
  >
    <View>
      <View
        pointerEvents="none"
        style={
          Array [
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 6,
            },
            Object {
              "backgroundColor": "#ccfeff",
              "borderColor": "rgba(0, 0, 0, 0.54)",
              "borderRadius": 4,
              "borderWidth": 1,
            },
          ]
        }
      />
      <View
        style={
          Object {
            "minHeight": 64,
            "paddingBottom": 0,
            "paddingTop": 8,
          }
        }
      >
        <View
          pointerEvents="none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "opacity": 1,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "transform": Array [
                Object {
                  "translateX": 0,
                },
              ],
            }
          }
        >
          <View
            pointerEvents="none"
            style={
              Object {
                "backgroundColor": "#ccfeff",
                "bottom": 0,
                "height": 2,
                "left": 10,
                "opacity": 0,
                "position": "absolute",
                "right": 0,
                "top": 6,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                ],
                "width": 8,
              }
            }
          />
          <Text
            numberOfLines={1}
            style={
              Object {
                "backgroundColor": "#ccfeff",
                "color": "transparent",
                "fontFamily": "System",
                "fontSize": 40,
                "fontWeight": "bold",
                "left": 18,
                "opacity": 0,
                "paddingHorizontal": 0,
                "position": "absolute",
                "textAlign": "left",
                "top": 37,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                  Object {
                    "translateY": 0,
                  },
                  Object {
                    "scale": 1,
                  },
                  Object {
                    "scaleY": 1,
                  },
                ],
                "writingDirection": "ltr",
              }
            }
          >
            CLIENT ID
          </Text>
          <Text
            numberOfLines={1}
            onLayout={[Function]}
            style={
              Object {
                "color": "#6200ee",
                "fontFamily": "System",
                "fontSize": 40,
                "fontWeight": "bold",
                "left": 0,
                "opacity": 0,
                "paddingHorizontal": 14,
                "position": "absolute",
                "textAlign": "left",
                "top": 36,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                  Object {
                    "translateY": 0,
                  },
                  Object {
                    "scale": 1,
                  },
                ],
                "writingDirection": "ltr",
              }
            }
          >
            CLIENT ID
          </Text>
          <Text
            numberOfLines={1}
            style={
              Object {
                "color": "rgba(0, 0, 0, 0.54)",
                "fontFamily": "System",
                "fontSize": 40,
                "fontWeight": "bold",
                "left": 0,
                "opacity": 0,
                "paddingHorizontal": 14,
                "position": "absolute",
                "textAlign": "left",
                "top": 36,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                  Object {
                    "translateY": 0,
                  },
                  Object {
                    "scale": 1,
                  },
                ],
                "writingDirection": "ltr",
              }
            }
          >
            CLIENT ID
          </Text>
        </View>
        <TextInput
          allowFontScaling={true}
          editable={true}
          multiline={false}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder=""
          placeholderTextColor="rgba(0, 0, 0, 0.54)"
          rejectResponderTermination={true}
          selectionColor="#6200ee"
          style={
            Array [
              Object {
                "flexGrow": 1,
                "margin": 0,
                "paddingHorizontal": 14,
                "textAlign": "left",
                "zIndex": 1,
              },
              Object {
                "height": 56,
              },
              Object {
                "paddingBottom": 0,
                "paddingTop": 0,
              },
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontSize": 40,
                "fontWeight": "bold",
                "textAlignVertical": "center",
              },
              Array [
                Object {},
              ],
            ]
          }
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
    </View>
  </View>
  <Text
    numberOfLines={1}
    style={
      Object {
        "alignSelf": "center",
        "backgroundColor": "#f50057",
        "borderRadius": 10,
        "color": "#ffffff",
        "fontFamily": "System",
        "fontSize": 40,
        "fontWeight": "400",
        "height": 53.5,
        "lineHeight": 20,
        "marginLeft": 220,
        "minWidth": 20,
        "opacity": 1,
        "overflow": "hidden",
        "paddingHorizontal": 4,
        "paddingTop": 13.5,
        "textAlign": "center",
        "textAlignVertical": "center",
        "width": 53.5,
      }
    }
  >
    0
  </Text>
  <Text
    style={
      Object {
        "color": "#2080aa",
        "fontSize": 53.5,
        "fontWeight": "bold",
        "padding": 10,
      }
    }
  >
    PASSWORD
  </Text>
  <View
    style={
      Object {
        "borderColor": "#2080aa",
        "borderRadius": 40,
        "borderWidth": 2,
        "marginBottom": 0,
        "padding": 10,
        "width": 300,
      }
    }
  >
    <View>
      <View
        pointerEvents="none"
        style={
          Array [
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 6,
            },
            Object {
              "backgroundColor": "#ccfeff",
              "borderColor": "rgba(0, 0, 0, 0.54)",
              "borderRadius": 4,
              "borderWidth": 1,
            },
          ]
        }
      />
      <View
        style={
          Object {
            "minHeight": 64,
            "paddingBottom": 0,
            "paddingTop": 8,
          }
        }
      >
        <View
          pointerEvents="none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "opacity": 1,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "transform": Array [
                Object {
                  "translateX": 0,
                },
              ],
            }
          }
        >
          <View
            pointerEvents="none"
            style={
              Object {
                "backgroundColor": "#ccfeff",
                "bottom": 0,
                "height": 2,
                "left": 10,
                "opacity": 0,
                "position": "absolute",
                "right": 0,
                "top": 6,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                ],
                "width": 8,
              }
            }
          />
          <Text
            numberOfLines={1}
            style={
              Object {
                "backgroundColor": "#ccfeff",
                "color": "transparent",
                "fontFamily": "System",
                "fontSize": 40,
                "fontWeight": "bold",
                "left": 18,
                "opacity": 0,
                "paddingHorizontal": 0,
                "position": "absolute",
                "textAlign": "left",
                "top": 37,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                  Object {
                    "translateY": 0,
                  },
                  Object {
                    "scale": 1,
                  },
                  Object {
                    "scaleY": 1,
                  },
                ],
                "writingDirection": "ltr",
              }
            }
          >
            PASSWORD
          </Text>
          <Text
            numberOfLines={1}
            onLayout={[Function]}
            style={
              Object {
                "color": "#6200ee",
                "fontFamily": "System",
                "fontSize": 40,
                "fontWeight": "bold",
                "left": 0,
                "opacity": 0,
                "paddingHorizontal": 14,
                "position": "absolute",
                "textAlign": "left",
                "top": 36,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                  Object {
                    "translateY": 0,
                  },
                  Object {
                    "scale": 1,
                  },
                ],
                "writingDirection": "ltr",
              }
            }
          >
            PASSWORD
          </Text>
          <Text
            numberOfLines={1}
            style={
              Object {
                "color": "rgba(0, 0, 0, 0.54)",
                "fontFamily": "System",
                "fontSize": 40,
                "fontWeight": "bold",
                "left": 0,
                "opacity": 0,
                "paddingHorizontal": 14,
                "position": "absolute",
                "textAlign": "left",
                "top": 36,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                  Object {
                    "translateY": 0,
                  },
                  Object {
                    "scale": 1,
                  },
                ],
                "writingDirection": "ltr",
              }
            }
          >
            PASSWORD
          </Text>
        </View>
        <TextInput
          allowFontScaling={true}
          editable={true}
          multiline={false}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder=""
          placeholderTextColor="rgba(0, 0, 0, 0.54)"
          rejectResponderTermination={true}
          selectionColor="#6200ee"
          style={
            Array [
              Object {
                "flexGrow": 1,
                "margin": 0,
                "paddingHorizontal": 14,
                "textAlign": "left",
                "zIndex": 1,
              },
              Object {
                "height": 56,
              },
              Object {
                "paddingBottom": 0,
                "paddingTop": 0,
              },
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontSize": 40,
                "fontWeight": "bold",
                "textAlignVertical": "center",
              },
              Array [
                Object {},
              ],
            ]
          }
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
    </View>
  </View>
  <Text
    numberOfLines={1}
    style={
      Object {
        "alignSelf": "center",
        "backgroundColor": "#f50057",
        "borderRadius": 10,
        "color": "#ffffff",
        "fontFamily": "System",
        "fontSize": 40,
        "fontWeight": "400",
        "height": 53.5,
        "lineHeight": 20,
        "marginLeft": 220,
        "minWidth": 20,
        "opacity": 1,
        "overflow": "hidden",
        "paddingHorizontal": 4,
        "paddingTop": 13.5,
        "textAlign": "center",
        "textAlignVertical": "center",
        "width": 53.5,
      }
    }
  >
    0
  </Text>
  <Text />
  <View
    accessibilityRole="progressbar"
    accessible={true}
    style={
      Array [
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        },
        undefined,
      ]
    }
  >
    <View
      collapsable={false}
      style={
        Object {
          "height": 48,
          "opacity": 1,
          "width": 48,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "bottom": 0,
            "justifyContent": "center",
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <View
          style={
            Object {
              "height": 48,
              "transform": Array [
                Object {
                  "rotate": "45deg",
                },
              ],
              "width": 48,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "height": 24,
                "overflow": "hidden",
                "width": 48,
              }
            }
          >
            <View
              style={
                Object {
                  "height": 48,
                  "transform": Array [
                    Object {
                      "translateY": 0,
                    },
                    Object {
                      "rotate": "-165deg",
                    },
                  ],
                  "width": 48,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "height": 24,
                    "overflow": "hidden",
                    "width": 48,
                  }
                }
              >
                <View
                  style={
                    Object {
                      "borderColor": "#f1c40f",
                      "borderRadius": 24,
                      "borderWidth": 4.8,
                      "height": 48,
                      "width": 48,
                    }
                  }
                />
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "bottom": 0,
            "justifyContent": "center",
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <View
          style={
            Object {
              "height": 48,
              "transform": Array [
                Object {
                  "rotate": "45deg",
                },
              ],
              "width": 48,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "height": 24,
                "overflow": "hidden",
                "top": 24,
                "width": 48,
              }
            }
          >
            <View
              style={
                Object {
                  "height": 48,
                  "transform": Array [
                    Object {
                      "translateY": -24,
                    },
                    Object {
                      "rotate": "345deg",
                    },
                  ],
                  "width": 48,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "height": 24,
                    "overflow": "hidden",
                    "width": 48,
                  }
                }
              >
                <View
                  style={
                    Object {
                      "borderColor": "#f1c40f",
                      "borderRadius": 24,
                      "borderWidth": 4.8,
                      "height": 48,
                      "width": 48,
                    }
                  }
                />
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
